File: debop4s-data/src/main/scala/com/github/debop4s/data/model/TreeNodePosition.scala
package com.github.debop4s.data.model

import com.github.debop4s.core.ValueObject
import com.github.debop4s.core.utils.{Hashs, ToStringHelper}
import javax.persistence.{Embeddable, Column}

/**
 * com.github.debop4s.data.model.TreeNodePosition
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2013. 12. 24. 오후 4:03
 */
@SerialVersionUID(3455568346636164669L)
@Embeddable
case class TreeNodePosition(var level: Int = 0, var order: Int = 0) extends ValueObject {

    @Column(name = "treeLevel")
    def getLevel: Int = level

    @Column(name = "treeOrder")
    def getOrder: Int = order

    def setPosition(level: Int, order: Int) {
        this.level = level
        this.order = order
    }

    override def hashCode(): Int = Hashs.compute(level, order)

    override protected def buildStringHelper: ToStringHelper =
        super.buildStringHelper
        .add("level", level)
        .add("order", order)
}

File: debop4s-data-orm/src/main/java/debop4s/data/orm/model/TreeNodePosition.java
package debop4s.data.orm.model;

import debop4s.core.ToStringHelper;
import debop4s.core.ValueObjectBase;
import debop4s.core.utils.Hashs;

import javax.persistence.Access;
import javax.persistence.AccessType;
import javax.persistence.Column;
import javax.persistence.Embeddable;

/**
 * 계층형 구조 엔티티의 TREE 상에서의 위치를 나타냅니다.
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 7. 11. 오전 9:02
 */
@Embeddable
@Access(AccessType.FIELD)
public class TreeNodePosition extends ValueObjectBase {

    public static TreeNodePosition copyFrom(TreeNodePosition src) {
        return new TreeNodePosition(src);
    }

    public TreeNodePosition() {
        this(0, 0);
    }

    public TreeNodePosition(Integer level, Integer order) {
        this.level = level;
        this.order = order;
    }

    public TreeNodePosition(TreeNodePosition src) {
        this.level = src.level;
        this.order = src.order;
    }

    public void setPosition(Integer level, Integer order) {
        this.level = level;
        this.order = order;
    }

    public void setPosition(TreeNodePosition src) {
        this.level = src.level;
        this.order = src.order;
    }

    /** 트리 레벨 (몇대손) */
    @Column(name = "treeLevel")
    private Integer level;

    /** 트리 노드 정렬 순서 - 형제끼리의 순서 */
    @Column(name = "treeOrder")
    private Integer order;

    public Integer getLevel() {
        return this.level;
    }

    public void setLevel(Integer level) {
        this.level = level;
    }

    public Integer getOrder() {
        return this.order;
    }

    public void setOrder(Integer order) {
        this.order = order;
    }

    @Override
    public int hashCode() {
        return Hashs.compute(level, order);
    }

    @Override
    public ToStringHelper buildStringHelper() {
        return super.buildStringHelper()
                    .add("level", level)
                    .add("order", order);
    }

    private static final long serialVersionUID = 3455568346636164669L;
}
